app/uk/gov/hmrc/agentregistration/shared/AgentApplication.scala

```diff
@@ -16,72 +16,117 @@
-    .applicantEmailAddress
-    .getOrThrowExpectedDataMissing(
-      "Applicant email address is not defined"
-    )
+  def isGrsDataReceived: Boolean =
+    applicationState match
+      case ApplicationState.Started => false
+      case ApplicationState.GrsDataReceived => true
+      case ApplicationState.Submitted => true
+
+  def getAmlsDetails: AmlsDetails = amlsDetails.getOrElse(expectedDataNotDefinedError("amlsDetails"))
+
+  private def as[T <: AgentApplication](using ct: reflect.ClassTag[T]): Option[T] =
+    this match
+      case t: T => Some(t)
+      case _ => None
+
+  private def asExpected[T <: AgentApplication](using ct: reflect.ClassTag[T]): T = as[T].getOrThrowExpectedDataMissing(
+    s"The application is not of the expected type. Expected: ${ct.runtimeClass.getSimpleName}, Got: ${this.getClass.getSimpleName}"
+  )
+
+  def asLlpApplication: AgentApplicationLlp = asExpected[AgentApplicationLlp]
+
+/** Sole Trader Application. This case class represents the data entered by a user for registering as a sole trader.
+  */
+final case class AgentApplicationSoleTrader(
+  override val internalUserId: InternalUserId,
+  override val groupId: GroupId,
+  override val createdAt: Instant,
+  override val applicationState: ApplicationState,
+  userRole: Option[UserRole] = None,
+  businessDetails: Option[BusinessDetailsSoleTrader],
+  override val amlsDetails: Option[AmlsDetails]
+)
+extends AgentApplication:
+
+  override val businessType: BusinessType.SoleTrader.type = BusinessType.SoleTrader
+  def getUserRole: UserRole = userRole.getOrElse(expectedDataNotDefinedError("userRole"))
+  def getBusinessDetails: BusinessDetailsSoleTrader = businessDetails.getOrElse(expectedDataNotDefinedError("businessDetails"))
+
+/** Application Applicatoin for Limited Liability Partnership (Llp). This case class represents the data entered by a user for registering as an Llp.
+  */
+final case class AgentApplicationLlp(
+  override val internalUserId: InternalUserId,
+  override val groupId: GroupId,
+  override val createdAt: Instant,
+  override val applicationState: ApplicationState,
+  businessDetails: Option[BusinessDetailsLlp],
+  applicantContactDetails: Option[ApplicantContactDetails],
+  override val amlsDetails: Option[AmlsDetails]
+)
+extends AgentApplication:
+
+  override val businessType: BusinessType.Partnership.LimitedLiabilityPartnership.type = BusinessType.Partnership.LimitedLiabilityPartnership
+
+  def getApplicantContactDetails: ApplicantContactDetails = applicantContactDetails.getOrThrowExpectedDataMissing("ApplicantContactDetails")
+  def getBusinessDetails: BusinessDetailsLlp = businessDetails.getOrThrowExpectedDataMissing("businessDetails")
+  def getCrn: Crn = getBusinessDetails.companyProfile.companyNumber
 
 object AgentApplication:
-  given format: OFormat[AgentApplication] = Json.format[AgentApplication]
+
+  @nowarn()
+  given OFormat[AgentApplication] =
```

Comment:
this shared model is missing the handle `format` as referenced by the codec in the back end repository
```suggestion
  given format: OFormat[AgentApplication] =
```